File: 3-serialize-btree/3-serialize-btree.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def serialize(node, arr):
	if not node:
		arr.append(-1)
		return
	arr.append(node.val)
	serialize(node.left, arr)
	serialize(node.right, arr)

def deserialize(arr):
	if arr[0] == -1:
		arr.pop(0)
		return None
	
	node = Node(arr[0])
	arr.pop(0)
	node.left = deserialize(arr)
	node.right = deserialize(arr)

	return node

def equal_tree(n1, n2):
	if not n1 and not n2:
		return True
	elif n1 and n2:
		return n1.val == n2.val and equal_tree(n1.left, n2.left) and equal_tree(n1.right, n2.right)
	return False

def main():
	node = Node(20)
	node.left = Node(11)
	node.right = Node(23)
	node.left.left = Node(8)
	node.left.right = Node(15)
	node.right.right = Node(28)
	node.right.right.right = Node(31)
	node.right.right.right.left = Node(29)
	node.right.right.right.right = Node(35)

	res = []
	serialize(node, res)
	print(res)

	node1 = deserialize(res)

	res = []
	serialize(node1, res)
	print(res)

	print(equal_tree(node, node1))

if __name__ == "__main__":
	main()

File: 3-serialize-btree/3-serialize-btree.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>
#include <iterator>
using namespace std;

struct Node {
    int key;
    struct Node *left, *right;
};

Node* makeNode(int key) 
{ 
    Node* temp = new Node; 
    temp->key = key; 
    temp->left = temp->right = NULL; 
    return temp; 
}

bool equalTree(Node *root1, Node *root2)
{
    if (root1 == NULL && root2 == NULL) {
        return true;
    } else if (root1 != NULL && root2 != NULL) {
        return root1->key == root2->key && 
            equalTree(root1->left, root2->left) && 
            equalTree(root1->right, root2->right);
    }
    return false;
}

void serialize(Node *root, string &acc)
{
    if (root == NULL) {
        acc.append("-1 ");
        return;
    }
    acc.append(to_string(root->key));
    acc.append(" ");
    serialize(root->left, acc);
    serialize(root->right, acc);
}

void deserialize(Node *&root, vector<int> &acc)
{
    if (acc[0] == -1) {
    	acc.erase(acc.begin());
    	return;
    }
    int val = acc[0];
    acc.erase(acc.begin());
    root = makeNode(val);
    deserialize(root->left, acc);
    deserialize(root->right, acc);
}

int main() 
{
    struct Node *root = makeNode(20);
	root->left = makeNode(11);
	root->right = makeNode(23);
	root->left->left = makeNode(8);
	root->left->right = makeNode(15);
	root->right->right = makeNode(28);
	root->right->right->right = makeNode(31);
	root->right->right->right->left = makeNode(29);
	root->right->right->right->right = makeNode(35);

    string s = "";
    serialize(root, s);
    cout << s << endl;

    istringstream iss(s);
    vector<int> tokens;
    copy(istream_iterator<int>(iss),
         istream_iterator<int>(),
         back_inserter(tokens));

    Node *node = NULL;
    deserialize(node, tokens);

    string s2 = "";
    serialize(node, s2);
    cout << s2 << endl;
    
    cout << equalTree(root, node) << endl;
}